default/AI/ResearchAI.py

```diff
@@ -20,19 +22,396 @@ def get_research_index():
+    most_adequate = 0
+    for specName in ColonisationAI.empire_colonizers:
+        environs = {}
+        this_spec = fo.getSpecies(specName)
+        if not this_spec:
+            continue
+        for ptype in [fo.planetType.swamp, fo.planetType.radiated, fo.planetType.toxic, fo.planetType.inferno, fo.planetType.barren, fo.planetType.tundra, fo.planetType.desert, fo.planetType.terran, fo.planetType.ocean, fo.planetType.asteroids]:
+            environ = this_spec.getPlanetEnvironment(ptype)
+            environs.setdefault(environ, []).append(ptype)
+        most_adequate = max(most_adequate, len(environs.get(fo.planetEnvironment.adequate, [])))
+    return most_adequate == 0
+
+def get_max_stealth_species():
+    stealth_grades = {'BAD': -15, 'GOOD': 15, 'GREAT': 40, 'ULTIMATE': 60}
+    stealth = -999
+    stealth_species = ""
+    for specName in ColonisationAI.empire_colonizers:
+        this_spec = fo.getSpecies(specName)
+        if not this_spec:
+            continue
+        this_stealth = stealth_grades.get(get_ai_tag_grade(list(this_spec.tags), "STEALTH"), 0)
+        if this_stealth > stealth:
+            stealth_species = specName
+            stealth = this_stealth
+    result = (stealth_species, stealth)
+    return result
+
+def get_ship_tech_usefulness(tech, ship_designer):
+    this_tech = fo.getTech(tech)
+    if not this_tech:
+        print "Invalid Tech specified"
+        return 0
+    unlocked_items = this_tech.unlockedItems
+    unlocked_hulls = []
+    unlocked_parts = []
+    for item in unlocked_items:
+        if item.type == fo.unlockableItemType.shipPart:
+            unlocked_parts.append(item.name)
+        elif item.type == fo.unlockableItemType.shipHull:
+            unlocked_hulls.append(item.name)
+    if not (unlocked_parts or unlocked_hulls):
+        return 0
+    old_designs = ship_designer.optimize_design(consider_fleet_count=False)
+    new_designs = ship_designer.optimize_design(additional_hulls=unlocked_hulls, additional_parts=unlocked_parts, consider_fleet_count=False)
+    if not (old_designs and new_designs):
+        # AI is likely defeated; don't bother with logging error message
+        return 0
+    old_rating, old_pid, old_design_id, old_cost = old_designs[0]
+    old_design = fo.getShipDesign(old_design_id)
+    old_rating = old_rating
+    new_rating, new_pid, new_design_id, new_cost = new_designs[0]
+    new_design = fo.getShipDesign(new_design_id)
+    new_rating = new_rating
+    if new_rating > old_rating:
+        ratio = (new_rating - old_rating) / (new_rating + old_rating)
+        return ratio * 1.5 + 0.5
+    else:
+        return 0
+
+def get_defense_priority(rng):
+    if foAI.foAIstate.aggression <= fo.aggression.cautious:
+        print "AI is cautious. Increasing priority for defense techs."
+        return 2
+    if foAI.foAIstate.misc.get('enemies_sighted', {}):
+        return 1
+    else:
+        return 0.2
+
+def get_production_boost_priority(rng):
+    return 1.5
```

Comment:
an industrialists check here is one of my changes as well, and I think your other points here are good as well.